publication-commons/src/test/java/no/unit/nva/doi/models/ImmutableDoiTest.java

```diff
@@ -0,0 +1,372 @@
+    }
+
+    @Test
+    void withPrefixWhenValueNotChangedReturnsSameInstance() {
+        var doi = createDoi();
+        assertThat(doi.withPrefix(EXAMPLE_PREFIX), is(sameInstance(doi)));
+    }
+
+    @Test
+    void withPrefixWhenValueHasChangedReturnsNewImmutableInstance() {
+        var doi = createDoi();
+        assertThat(doi.withPrefix(EXAMPLE_PREFIX + EXAMPLE_RANDOM_VALUE), not(is(sameInstance(doi))));
+    }
+
+    @Test
+    void withSuffixWhenValueNotChangedReturnsSameInstance() {
+        var doi = createDoi();
+        assertThat(doi.withSuffix(EXAMPLE_SUFFIX), is(sameInstance(doi)));
+    }
+
+    @Test
+    void withSuffixWhenValueHasChangedReturnsNewImmutalbeInstance() {
+        var doi = createDoi();
+        assertThat(doi.withSuffix(EXAMPLE_SUFFIX + EXAMPLE_RANDOM_VALUE), not(is(sameInstance(doi))));
+    }
+
+    @Test
+    void hashCodeReturnsSameHashCodeForTwoDifferentInstancesButIdenticalObject() {
+        var doi = createDoi();
+        var doi2 = createDoi();
+
+        assertThat(doi.hashCode(), is(equalTo(doi2.hashCode())));
+    }
+
+    @Test
+    void equalsReturnsTrueForTwoDifferentInstancesButIdenticalObjects() {
+        var doi = createDoi();
+        var doi2 = createDoi();
+
+        assertThat(doi, is(equalTo(doi2)));
+        assertThat(doi, is(equalTo(doi)));
+    }
+
+    @Test
+    void toIdThenReturnsDoiIdWhichIsProxyPrefixAndSuffix() {
+        String randomSuffix = createRandomSuffix();
+        var doi = createDoi(randomSuffix);
+        URI expectedUri = URI.create(EXAMPLE_PROXY + EXAMPLE_PREFIX + FORWARD_SLASH + randomSuffix);
+        assertThat(doi.toUri(), is(equalTo(expectedUri)));
+    }
+
+    @Test
+    void toIdWhereBuilderWithProxyWithoutSuffixSlashReturnsCorrectIdURI() {
+        var randomSuffix = createRandomSuffix();
+        var doi = ImmutableDoi.builder()
+            .withProxy(URI.create("https://doi.org"))
+            .withPrefix(DEMO_PREFIX)
+            .withSuffix(randomSuffix)
+            .build();
+        assertThat(doi.toUri(),
+            is(equalTo(URI.create(DEFAULT_DOI_PROXY + DEMO_PREFIX + FORWARD_SLASH + randomSuffix))));
+    }
+
+    @Test
+    void builderWithIdentifierSetsPrefixAndSuffixCorrect() {
+        Doi doi = ImmutableDoi.builder().withIdentifier(EXAMPLE_IDENTIFIER).build();
+        assertThat(doi.getPrefix(), is(equalTo(EXAMPLE_PREFIX)));
+        assertThat(doi.getSuffix(), is(equalTo(EXAMPLE_SUFFIX)));
+    }
+
+    @Test
+    void toIdWithAnotherSubClassOfDoiWithInvalidProxyUriThenThrowsIllegalStateException() {
+        var doi = getAnotherPojoDoi(URI.create(URI_VALID_EMAILTO_BUT_INVALID_URL));
+        var actualException = assertThrows(IllegalStateException.class, doi::toUri);
+        assertThat(actualException.getMessage(), is(equalTo(Doi.ERROR_PROXY_URI_MUST_BE_A_VALID_URL)));
+    }
+
+    @Test
+    void withProxyWithInvalidUrlAsUriThrowsIllegalStateException() {
```

Comment:
remove these methods and use the pattern. I do not think the performance improvements on copying dois are critical. Besides, do we have such an example where we create a DOI from another DOI? 
```
doi.copy().withBlabla().build()
public Builder copy(){

}

Reply:
Removed.